File: prototype/sim/ai.js
// AI - a port of BaseAIController.DoAIStuff (docs/DESIGN.md §5):
// one decision procedure per turn issuing the same orders a player would.
(function (global) {
  "use strict";
  const isNode = typeof module !== "undefined" && module.exports;
  const dmath = isNode ? require("./dmath.js") : global.FT.dmath;
  const data = isNode ? require("./data.js") : global.FT.data;
  const { V } = dmath;
  const { SHIP_CLASSES } = data;

  function planShip(state, ship, rng) {
    const cls = SHIP_CLASSES[ship.classKey];
    const enemies = state.ships.filter(s => !s.destroyed && s.faction !== ship.faction);
    if (enemies.length === 0) return { move: { mode: "MOVE_AND_TURN" } };

    // target: keep retaliation target if alive, else FIRST live enemy
    // (registration order - kill priority is emergent, like the original).
    //
    // The choice is REPORTED in the order rather than written onto the ship
    // here. Planning must not mutate state: a client replaying a stored turn,
    // or receiving these orders over the wire, never calls planShip, so a
    // side effect written here would exist on one machine and not the other
    // and the state hashes would diverge (ADR-6). resolveTurn applies
    // order.aiTarget on every path instead.
    let target = ship.ai.targetId ? enemies.find(e => e.id === ship.ai.targetId) : null;
    if (!target) target = enemies[0];

    // chase boost: if target beyond range, AI gets extra reach (the AIBoost cheat)
    const dist = V.dist(ship.pos, target.pos);
    let range = cls.thrusterRange;
    if (ship.ai.canChase && dist > range) range = range + 20;

    // destination: random point on a sphere around the target (25%..100% of range)
    const radial = rng.onUnitSphere(V);
    const orbit = rng.range(0.25 * range, range);
    const dest = V.add(target.pos, V.scale(radial, orbit));

    // face the target's current position (snap-rotation approximation)
    const face = V.norm(V.sub(target.pos, ship.pos));

    // weapons: one random second 1..8; below fireProbability 0.2 the AI
    // deterministically queues only weapon[0] (the shipped low-aggression branch)
    const second = rng.int(1, 9);
    const weapons = [];
    for (let i = 0; i < ship.weapons.length; i++) {
      if (ship.ai.fireProbability < 0.2) {
        if (i === 0) weapons.push({ weaponIndex: 0, second, targetShipId: target.id });
      } else if (rng.float() < ship.ai.fireProbability) {
        weapons.push({ weaponIndex: i, second, targetShipId: target.id });
      }
    }

    // boarding-specialist behavior: rogue-class AI boards when close and healthy
    let board;
    if (cls.boardingRange >= 40 && ship.marines > cls.boardingCapacity && dist <= cls.boardingRange) {
      board = target.id;
    }

    return {
      move: { mode: "MOVE_AND_TURN", target: [dest.x, dest.y, dest.z], face: [face.x, face.y, face.z] },
      weapons,
      board,
      aiTarget: target.id,
    };
  }

  const api = { planShip };
  global.FT = global.FT || {};
  global.FT.ai = api;
  if (typeof module !== "undefined" && module.exports) module.exports = api;
})(typeof globalThis !== "undefined" ? globalThis : this);

File: engine/sim_core/src/ai.rs
//! The AI, a port of the archive's `BaseAIController.DoAIStuff`
//! (docs/DESIGN.md section 5): one decision procedure per turn, issuing the
//! same orders a player would issue.
//!
//! It reads state and returns an order. It writes nothing. That is not
//! tidiness: a client replaying a stored turn, or receiving orders over the
//! wire, never calls this, so anything written onto a ship here would exist on
//! one machine and not the other and the state hashes would part several turns
//! later with no clue why (ADR-6). The target choice is REPORTED in the order
//! and applied by the resolver on every path instead.

use crate::flight::Mode;
use crate::state::Sim;
use crate::turn::{FireOrder, Order};

pub fn plan_ship(sim: &Sim, si: usize) -> Order {
    let ship = &sim.ships[si];
    let cls = ship.class_def();

    let enemies: Vec<usize> = sim
        .ships
        .iter()
        .enumerate()
        .filter(|(_, s)| !s.destroyed && s.faction != ship.faction)
        .map(|(i, _)| i)
        .collect();
    if enemies.is_empty() {
        return Order { mode: Some(Mode::MoveAndTurn), ..Default::default() };
    }

    let mut rng = crate::rng::Rng::new(&sim.seed, sim.turn, &format!("ai:{}", si));

    // Keep the retaliation target while it lives, else take the first live
    // enemy by index. Kill priority is emergent from registration order, the
    // same way the archive's was, rather than being a rule anyone wrote.
    let ti = ship
        .ai_target
        .map(|t| t as usize)
        .filter(|t| enemies.contains(t))
        .unwrap_or(enemies[0]);
    let target = &sim.ships[ti];

    let dist = ship.pos.dist(target.pos);
    let mut range = ship.flight.nominal_reach();
    // The chase cheat: an AI that cannot close would just sit there.
    if ship.ai_can_chase && dist > range {
        range += 20.0;
    }

    // Destination: a point on a sphere around the target, somewhere between a
    // quarter and all of the engagement range.
    let radial = rng.on_unit_sphere();
    let orbit = rng.range(0.25 * range as f64, range as f64) as f32;
    let dest = target.pos.add(radial.scale(orbit));
    let face = target.pos.sub(ship.pos).norm();

    // One random second, 1..=8, for every weapon that fires this turn. Below a
    // fire probability of 0.2 the AI deterministically queues only its first
    // weapon, which is the low aggression branch the archive shipped.
    let second = rng.int(1, 9);
    let mut weapons = Vec::new();
    for i in 0..ship.weapons.len() {
        if ship.ai_fire_probability < 0.2 {
            if i == 0 {
                weapons.push(FireOrder {
                    weapon_index: 0,
                    second,
                    target_ship: target.id,
                    target_sub: None,
                });
            }
        } else if (rng.float() as f32) < ship.ai_fire_probability {
            weapons.push(FireOrder {
                weapon_index: i,
                second,
                target_ship: target.id,
                target_sub: None,
            });
        }
    }

    // The boarding specialist actually specialises: a rogue hull closes and
    // boards when it is near enough and still has marines to spare.
    let board = if cls.boarding_range >= 40.0
        && ship.marines > cls.boarding_capacity
        && dist <= cls.boarding_range
    {
        Some(target.id)
    } else {
        None
    };

    Order {
        mode: Some(Mode::MoveAndTurn),
        target: Some(dest),
        face: Some(face),
        weapons,
        board,
        ai_target: Some(target.id),
    }
}
